doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1208 @@
+One of these ``CastingImpl`` objects is special because it should define
+the cast within the same DType (from one instance to another).
+A DType which does not define this, must have only a single implementation
+and not be parametric.
+
+DTypes such as a Unit datatype, wrapping an existing Numerical datatype *must*
+be able to access these ``CastingImpl``.
+They will be able to wrap and modify them as necessary. This means that
+a DType which wraps another can automatically define casts for any DType it
+knows at construction time (see also Alternatives).
+
+These return a ``CastingImpl`` defined in some more detail in the next section.
+It also answers the last question: ``np.can_cast(DType, OtherDType, "safe")``
+since ``CastingImpl`` defines a ``CastingImpl.cast_kind = "safe"``.
+
+Each ``CastingImpl`` has a specific DType signature:
+``CastingImpl[InputDtype, RequestedDtype]``.
+Additionally, it will have one more method::
+
+    adjust_descriptors(self, Tuple[DType] : input) -> casting, Tuple[DType]
+
+(this method is common with the ufunc machinery, see NEP 43).
+Here, valid values for ``input`` are:
+
+* ``(input_dtype, None)``
+* ``(input_dtype, requested_dtype)``
+
+with the correct DType classes.
+In the first case, when ``None`` is given, no dtype instance was requested.
+The returned values will be a new tuple of two datatypes, filling in ``None``
+if necessary.
+Note that the output *can* differ from the input.
+At the beginning casting was described as a, possibly, three step process.
+The ``CastingImpl`` only defines the single step 2.
+If the returned datatypes differ from the provided ones, this means that additional
+casting steps are required for either input or output.
+
+The casting returned by ``adjust_descriptors`` and set on the ``CastingImpl`` as
+a maximum version should consist of safest cast which is allowed:
+
+* ``NPY_SAFE_CASTING``, ``NPY_SAME_KIND_CASTING``, ``NPY_UNSAFE_CASTING``
+* a ``NPY_CAST_IS_VIEW`` *flag*
+
+Where the current ``NPY_EQUIV_CASTING`` should be signaled by
+``NPY_SAFE_CASTING | NPY_CAST_IS_VIEW``. By returning this, the error
+message is given in a place with more information available, and new modes
+could be added in principle.
+(The general casting does not need to define ``CAST_IS_VIEW``.)
+
+**Casting between instances of the same DType:**
+
+One of the casting implementations added should define ``CastingImpl[DType, DType]``,
+this implementation must be capable of handling any remaining steps, such
+as byteswapping.
+If not implemented, the datatype may only have a singleton implementation.
+
+**Actual casting:**
+
+To provide the actual casting functionality, an additional method:
+
+* ``get_transferfunction(...)``
+
+is necessary to provide a low-level C-implementation.
+However, this method shall *not* be part of the public API, instead
+users will initially be limited in what casting functions they can provide
+(e.g. only contiguous loops of multiple items as is used right now).
```

Comment:
I am currently refusing to do such complex casts. A DType can only define a cast from one of itself to another.

I had currently no intention of "finding a path".  Such direct cast would be either undefined, _or_ one of the DTypes would have to hard-code such a path/fallback.  I would hope that there would be no issue with adding such things later, although I also somewhat hope that it is simply not actually necessary here and would just be over-engineering :).

Reply:
@seberg - definitely fair to keep it simple, and I think as long as there is a general function that does the "search" for a transfer function, it would be extensible anyway.

Still, it may be good to discuss explicitly the straw-man `int24` type that has come up before. If a user were to define it, would there need to be definitions of conversion functions to all numpy `int` and `uint` (and float) types, or just a generic conversion from `int` (and `float`?) and perhaps fast ones from `np.int16` and `np.int8` or so (ie., would it be possible to have conversion if the other `dtype` provided a way to convert to a type in common, say `int` or `float`).

Note that I don't think this actually affects anything for the new `dtype` class, and perhaps it is discussed already in one of the other NEPs... If so, all that's needed here might be a cross-reference.

Reply:
Yes, as of my thoughts now `int24` would have to define all of those casts manually.  In theory it could be simple to extend this I guess.
One option could be to allow a CastingImpl (and generally cast) be a multi-step (instead of at most 3 step) process.  You would still have to define a `CastingImpl[int8, int24]`, but internally, it could just point to the same loop as `CastingImpl[int32, int24]`, and its `adjust_descriptor` function would return `int32, int24` instead of `int8,int24`. At which point NumPy would know that it has to do more steps to do the cast.

That might be a bit of a middle ground.  You still need to *define* all the casts, but it can be sufficient to write fewer transfer functions.  If buggy, such a chain of casts may never settle I suppose, but in practice I guess it should be OK and likely not even much more difficult.

I think it is important to note that this is at least a very much possible extension of the API.  The alternative is to have the API getting a full cast-function to actually try to find the graph by itself.  But at that point even the question of whether or not a cast is safe would be a multi-step process.

Reply:
OK, thanks. I indeed think it is extensible, so for now maybe all that is needed is to be explicit about what it is now and that it is possible to extend if the need arises. Clearly, it will never *hurt* to define all the transformations explicitly!